File: lib/cusip_style_luhn_mod_n.rb
require "cusip_style_luhn_mod_n/version"

module CUSIPStyleLuhnModN
  def self.valid?(str)
    str[-1] == check_digit(str[0...-1]).to_s
  end

  def self.check_digit(s)
    total = 0
    s.split('').reverse.each_with_index do |c, i|
      value = character_value(c)
      multiplier = (i % 2).zero? ? 2 : 1
      sum_of_digits = digits(value * multiplier).reduce(:+)
      total += sum_of_digits
    end
    total %= 10
    total.zero? ? 0 : 10 - total
  end

  def self.character_value(c)
    value = c.ord - '0'.ord
    return value if value >= 0 && value <= 9
    value = c.downcase.ord - 'a'.ord + 10
    return value if value >= 0 && value <= 35
    raise "Illegal character: #{c}"
  end

  def self.digits(n)
    digits = []
    return [0] if n.zero?
    while n != 0
      n, last_digit = n.divmod(10)
      digits << last_digit
    end
    digits.reverse
  end
end

File: javascripts/cusip_style_luhn_mod_n.js
(function(global) {
  var CUSIPStyleLuhnModN = {};

  CUSIPStyleLuhnModN.isValid = function(str) {
    return str.slice(-1) === CUSIPStyleLuhnModN.checkDigit(str.slice(0, -1)).toString();
  };

  CUSIPStyleLuhnModN.checkDigit = function(s) {
    var total = 0;
    s.split("").reverse().forEach(function (c, i) {
      var value = characterValue(c);
      var multiplier = (i % 2 === 0) ? 2 : 1;
      total += digits(value * multiplier).reduce(function (prev, current) {
          return prev + current;
      }, 0);
    });
    total %= 10;
    return total === 0 ? 0 : 10 - total;
  };

  function characterValue(c) {
    var value = c.toLowerCase().charCodeAt(0);
    var digitValue = value - "0".charCodeAt(0);
    if (digitValue >= 0 && digitValue <= 9) {
      return digitValue;
    }
    var letterValue = value - "a".charCodeAt(0);
    if (letterValue >= 0 && letterValue <= 35) {
      return letterValue + 10;
    }
    throw "Illegal character: " + c;
  };

  function digits(n) {
    if (n === 0) {
      return [0];
    }
    var digits = [];
    while (n !== 0) {
      digits.unshift(n % 10);
      n = (n / 10 >> 0);
    }
    return digits;
  }

  if (typeof exports !== 'undefined') {
    if (typeof module !== 'undefined' && module.exports) {
      exports = module.exports = CUSIPStyleLuhnModN;
    }
    exports.CUSIPStyleLuhnModN = CUSIPStyleLuhnModN;
  }
  else {
    global.CUSIPStyleLuhnModN = CUSIPStyleLuhnModN;
  }
})(this);
